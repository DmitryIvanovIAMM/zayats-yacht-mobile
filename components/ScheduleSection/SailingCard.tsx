import { primary, secondary } from "@/constants/Colors";
import { getInternationalDateFormat } from "@/helpers/dateTime";
import {
  calculateDaysInTransit,
  calculateMilesForRoute
} from "@/helpers/routesCalculators";
import { PortFrontend } from "@/models/PortFrontend";
import { ShipStopWithSailingAndPortFrontend } from "@/models/ShipStopFrontend";
import React from "react";
import { Platform, StyleSheet, View } from "react-native";
import { Button, Card, Divider, Text } from "react-native-paper";

interface SailingCardProps {
  route: ShipStopWithSailingAndPortFrontend[];
}

export const SailingCard: React.FC<SailingCardProps> = ({
  route
}: SailingCardProps) => {
  return (
    <Card style={styles.card}>
      <Card.Cover
        source={{
          //https://zayats-yacht.vercel.app/_next/image?url=%2Fimages%2FGolfito.jpg&w=3840&q=75
          uri: `https://zayats-yacht.vercel.app/_next/image?url=%2Fimages%2F${(route[route.length - 1].departurePort as PortFrontend).imageFileName}&w=3840&q=75`
        }}
        style={{ height: 200, width: "100%", borderRadius: 0 }}
      />
      <Card.Content>
        <Text style={styles.title}>
          {route[route.length - 1].sailing?.name || ""}
        </Text>
        <View style={styles.descriptionContainer}>
          <View style={styles.descriptionContainerColumn}>
            <Text style={styles.columnHeader}>Loading Port</Text>
          </View>
          <View style={styles.descriptionContainerColumn}>
            <Text style={styles.columnContent}>
              {route[0].departurePort?.portName}
            </Text>
          </View>
        </View>
        <View style={styles.descriptionContainer}>
          <View style={styles.descriptionContainerColumn}>
            <Text style={styles.columnHeader}>Destination Port</Text>
          </View>
          <View style={styles.descriptionContainerColumn}>
            <Text style={styles.columnContent}>
              {route[route.length - 1].departurePort?.portName || ""}
            </Text>
          </View>
        </View>
        <View style={styles.descriptionContainer}>
          <View style={styles.descriptionContainerColumn}>
            <Text style={styles.columnHeader}>Loading Date</Text>
          </View>
          <View style={styles.descriptionContainerColumn}>
            <Text style={styles.columnContent}>
              {getInternationalDateFormat(route[0].arrivalOn)}
            </Text>
          </View>
        </View>
        <View style={styles.descriptionContainer}>
          <View style={styles.descriptionContainerColumn}>
            <Text style={styles.columnHeader}>Arrival Date</Text>
          </View>
          <View style={styles.descriptionContainerColumn}>
            <Text style={styles.columnContent}>
              {getInternationalDateFormat(route[route.length - 1].arrivalOn)}
            </Text>
          </View>
        </View>
        <Divider style={styles.divider} />
        <View style={styles.integrationContainer}>
          <Text style={styles.integrationHeader}>Miles:</Text>
          <Text style={styles.integrationText}>
            {calculateMilesForRoute(route)}
          </Text>
        </View>
        <View style={styles.integrationContainer}>
          <Text style={styles.integrationHeader}>Days:</Text>
          <Text style={styles.integrationText}>
            {calculateDaysInTransit(route)} days
          </Text>
        </View>
      </Card.Content>
      <Card.Actions>
        <Button style={styles.button}>
          <Text style={styles.buttonText}>Get Quote</Text>
        </Button>
      </Card.Actions>
    </Card>
  );
};

const styles = StyleSheet.create({
  card: Platform.select({
    web: {
      backgroundColor: "#fff",
      borderRadius: 0,
      padding: 0,
      minHeight: 194,
      margin: 20,
      boxShadow: "0 12px 48px 0 rgba(0,0,0,0.38), 0 6px 24px 0 rgba(0,0,0,0.32)"
    },
    default: {
      backgroundColor: "#fff",
      borderRadius: 0,
      padding: 0,
      minHeight: 194,
      margin: 20,
      shadowColor: "#000",
      shadowOpacity: 0.3,
      shadowRadius: 6,
      elevation: 3
    }
  }),
  title: Platform.select({
    default: {
      fontSize: 20,
      fontWeight: "bold",
      marginBottom: 4
    },
    android: {
      fontSize: 16,
      color: "black",
      fontWeight: "bold",
      marginBottom: 4
    }
  }),
  descriptionContainer: Platform.select({
    default: {
      marginTop: 10,
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between"
    },
    android: {
      marginTop: 0,
      display: "flex",
      flexDirection: "row",
      justifyContent: "space-between"
    }
  }),
  descriptionContainerColumn: {
    flex: 1,
    justifyContent: "flex-start"
  },
  columnHeader: Platform.select({
    default: {
      fontSize: 16,
      color: secondary.dark
    },
    android: {
      fontSize: 14,
      color: secondary.dark
    }
  }),
  columnContent: Platform.select({
    default: {
      fontSize: 16,
      fontWeight: "bold",
      color: primary.dark
    },
    android: {
      fontSize: 14,
      fontWeight: "bold",
      color: primary.dark
    }
  }),
  divider: {
    marginVertical: 10
  },
  integrationContainer: {
    marginTop: 0,
    display: "flex",
    flexDirection: "row",
    alignItems: "center"
  },
  integrationHeader: Platform.select({
    default: {
      fontSize: 16,
      color: secondary.dark,
      marginRight: 10,
      textTransform: "uppercase"
    },
    android: {
      fontSize: 14,
      color: secondary.dark,
      marginRight: 10,
      textTransform: "uppercase"
    }
  }),
  integrationText: Platform.select({
    default: {
      fontSize: 16,
      fontWeight: "bold",
      color: secondary.dark
    },
    android: {
      fontSize: 14,
      fontWeight: "bold",
      color: secondary.dark
    }
  }),
  button: {
    marginTop: 0,
    padding: 0,
    color: "white",
    height: 40,
    borderWidth: 0,
    backgroundColor: secondary.main,
    borderRadius: 0,
    width: "100%"
  },
  buttonText: {
    color: "white",
    fontSize: 16,
    fontWeight: "bold",
    textTransform: "uppercase",
    textAlign: "center"
  }
});
