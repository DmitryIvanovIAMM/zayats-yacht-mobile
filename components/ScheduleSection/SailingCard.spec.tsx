import { render } from "@testing-library/react-native";
import React from "react";
import { SailingCard } from "./SailingCard"; // changed to named import

// Mock helpers used inside the component
jest.mock("@/helpers/dateTime", () => {
  return {
    getInternationalDateFormat: (date: any) => `DATE:${String(date)}`,
  };
});

jest.mock("@/helpers/routesCalculators", () => {
  return {
    calculateMilesForRoute: () => 1234,
    calculateDaysInTransit: () => 7,
  };
});

// Mock react-native-paper to simple primitives (avoid Provider/native deps)
jest.mock("react-native-paper", () => {
  const React = require("react");
  const { View, Text, Image } = require("react-native");

  const Card = ({ children, ...props }: any) => (
    <View {...props}>{children}</View>
  );
  (Card as any).Cover = ({ source, ...props }: any) =>
    React.createElement(Image, { source, ...props });
  (Card as any).Content = ({ children, ...props }: any) =>
    React.createElement(View, { ...props }, children);
  (Card as any).Actions = ({ children, ...props }: any) =>
    React.createElement(View, { ...props }, children);

  const Button = ({ children, ...props }: any) =>
    React.createElement(
      View,
      { accessibilityRole: "button", ...props },
      children
    );
  const Divider = (props: any) => React.createElement(View, props);

  return { Card, Button, Divider, Text };
});

describe("SailingCard", () => {
  const route = [
    {
      arrivalOn: "2025-01-01",
      departurePort: { portName: "Port A" },
    },
    {
      arrivalOn: "2025-02-02",
      departurePort: { portName: "Port B", imageFileName: "Golfito.jpg" },
      sailing: { name: "Trans-Atlantic" },
    },
  ] as any;

  it("renders core info (title, ports, dates, metrics, button)", () => {
    const { getByText } = render(<SailingCard route={route} />);

    // Title (sailing name)
    expect(getByText("Trans-Atlantic")).toBeTruthy();

    // Ports
    expect(getByText("Loading Port")).toBeTruthy();
    expect(getByText("Port A")).toBeTruthy();
    expect(getByText("Destination Port")).toBeTruthy();
    expect(getByText("Port B")).toBeTruthy();

    // Dates (formatted by mocked formatter)
    expect(getByText("Loading Date")).toBeTruthy();
    expect(getByText("DATE:2025-01-01")).toBeTruthy();
    expect(getByText("Arrival Date")).toBeTruthy();
    expect(getByText("DATE:2025-02-02")).toBeTruthy();

    // Metrics (from mocked calculators)
    expect(getByText("Miles:")).toBeTruthy();
    expect(getByText("1234")).toBeTruthy();
    expect(getByText("Days:")).toBeTruthy();
    expect(getByText("7 days")).toBeTruthy();

    // Button text
    expect(getByText("Get Quote")).toBeTruthy();
  });

  it("uses last port image in Card.Cover URI", () => {
    const { toJSON } = render(<SailingCard route={route} />);
    const tree = toJSON();

    // Walk the rendered tree to find Image node with 'source.uri'
    const findImageWithUri = (node: any): string | null => {
      if (!node) return null;
      if (node.type === "Image" && node.props?.source?.uri) {
        return node.props.source.uri as string;
      }
      const children = node.children || [];
      for (const child of children) {
        if (typeof child === "object" && child) {
          const res = findImageWithUri(child);
          if (res) return res;
        }
      }
      return null;
    };

    const uri = findImageWithUri(tree);
    expect(uri).toBeTruthy();

    // Support both raw path and Next.js optimized URL with encoded query param
    let decoded = uri as string;
    try {
      const u = new URL(decoded);
      const qp = u.searchParams.get("url");
      if (qp) decoded = decodeURIComponent(qp);
    } catch {
      // not a full URL; keep original
    }

    expect(decoded).toContain("/images/Golfito.jpg");
  });

  it("matches snapshot", () => {
    const { toJSON } = render(<SailingCard route={route} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
