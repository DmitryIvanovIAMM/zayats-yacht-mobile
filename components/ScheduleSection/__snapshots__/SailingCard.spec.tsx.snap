// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SailingCard matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#fff",
      "borderRadius": 0,
      "elevation": 3,
      "margin": 20,
      "minHeight": 194,
      "padding": 0,
      "shadowColor": "#000",
      "shadowOpacity": 0.3,
      "shadowRadius": 6,
    }
  }
>
  <Image
    source={
      {
        "uri": "https://zayats-yacht.vercel.app/_next/image?url=%2Fimages%2FGolfito.jpg&w=3840&q=75",
      }
    }
    style={
      {
        "borderRadius": 0,
        "height": 200,
        "width": "100%",
      }
    }
  />
  <View>
    <Text
      style={
        {
          "color": "#006666",
          "fontSize": 20,
          "fontWeight": "bold",
          "marginBottom": 4,
        }
      }
    >
      Trans-Atlantic
    </Text>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 10,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            {
              "color": "#006666",
              "fontSize": 16,
            }
          }
        >
          Loading Port
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            {
              "color": "#3365A7",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Port A
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 10,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            {
              "color": "#006666",
              "fontSize": 16,
            }
          }
        >
          Destination Port
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            {
              "color": "#3365A7",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          Port B
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 10,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            {
              "color": "#006666",
              "fontSize": 16,
            }
          }
        >
          Loading Date
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            {
              "color": "#3365A7",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          DATE:2025-01-01
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 10,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            {
              "color": "#006666",
              "fontSize": 16,
            }
          }
        >
          Arrival Date
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
            "justifyContent": "flex-start",
          }
        }
      >
        <Text
          style={
            {
              "color": "#3365A7",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          DATE:2025-02-02
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "marginVertical": 10,
        }
      }
    />
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "marginTop": 0,
        }
      }
    >
      <Text
        style={
          {
            "color": "#006666",
            "fontSize": 16,
            "marginRight": 10,
            "textTransform": "uppercase",
          }
        }
      >
        Miles:
      </Text>
      <Text
        style={
          {
            "color": "#006666",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        1234
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "marginTop": 0,
        }
      }
    >
      <Text
        style={
          {
            "color": "#006666",
            "fontSize": 16,
            "marginRight": 10,
            "textTransform": "uppercase",
          }
        }
      >
        Days:
      </Text>
      <Text
        style={
          {
            "color": "#006666",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        7
         days
      </Text>
    </View>
  </View>
  <View>
    <View
      accessibilityRole="button"
      style={
        {
          "backgroundColor": "#339966",
          "borderRadius": 0,
          "borderWidth": 0,
          "color": "white",
          "height": 40,
          "marginTop": 0,
          "padding": 0,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
            "textAlign": "center",
            "textTransform": "uppercase",
          }
        }
      >
        Get Quote
      </Text>
    </View>
  </View>
</View>
`;
